swarm/storage/localstore/subscription_pull.go

```diff
@@ -135,7 +132,12 @@ func (db *DB) SubscribePull(ctx context.Context, bin uint8, since, until uint64)
 					log.Error("localstore pull subscription iteration", "bin", bin, "since", since, "until", until, "err", err)
 					return
 				}
-				first = false
+
+				// if we sent some chunks on the channel - return so that the consumer can
+				// close the batch
+				if count > 0 {
```

Comment:
I am pretty sure `SubscribePull` and `SubscribePush` are not meant to be used in this way.

You are not supposed to exit the forever goroutine because you have added chunks.

AFAIK `SubscribePull` and `SubscribePush` return a channel that you read for incoming chunks and consumers or producers can decide when to terminate and signal to do that.

Returning here would also create resource leaks I think.